src/Microsoft.AspNetCore.Routing/EndpointRoutingMiddleware.cs

```diff
@@ -55,18 +55,19 @@ internal sealed class EndpointRoutingMiddleware
 
         public async Task Invoke(HttpContext httpContext)
         {
-            // For back-compat EndpointRouteValuesFeature implements IEndpointFeature, IRouteValuesFeature and IRoutingFeature
             var feature = new EndpointFeature();
 
-            SetEndpointFeature(httpContext, feature);
-
             // There's an inherent race condition between waiting for init and accessing the matcher
             // this is OK because once `_matcher` is initialized, it will not be set to null again.
             var matcher = await InitializeAsync();
 
             await matcher.MatchAsync(httpContext, feature);
             if (feature.Endpoint != null)
             {
+                // Set the endpoint feature only on success. This means we won't overwrite any
+                // existing state for related features unless we did something.
+                SetEndpointFeature(httpContext, feature);
```

Comment:
@davidfowl @Tratcher - what's the right thing to do here. Is it good or bad to avoid setting the feature if we don't match?

Reply:
Ping. Do you guys have an opinion on this?

Reply:
I think it makes sense.

Reply:
From a backwards compat perspective, what does the current routing do? Will it always set IRoutingFeature? Does it matter if that behavior changes?

Reply:
Current routing middleware only sets the feature if it found something (and executes it immediately). It's not super meaningful outside of MVC because old routing is terminal.

Reply:
What happens if UseMvc doesn't match anything? Does it throw/return then or does the host return 404 at the end?

Reply:
UseMvc just yields control the next middleware

Reply:
Great. Might be worth having a functional test of middleware that follows UseMvc to check compatibility with endpoint routing.

Reply:
If you did set the feature for a non-match, what would it contain? Nulls? Then you're better off not setting the feature.

Behaviorally all feature interfaces should be treated as optional except for request and response.